unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatHUDController.cs

```diff
@@ -109,29 +118,131 @@ public void SetVisibility(bool visible)
+
+        model.messageId = message.messageId;
+        model.messageType = message.messageType;
+        model.bodyText = message.body;
+        model.timestamp = message.timestamp;
+
+        if (!string.IsNullOrEmpty(message.recipient))
+        {
+            model.isChannelMessage = chatController.GetAllocatedChannel(message.recipient) != null;
+
+            if (!model.isChannelMessage)
+            {
+                UserProfile recipientProfile = userProfileBridge.Get(message.recipient);
+
+                if (recipientProfile != null)
+                    model.recipientName = recipientProfile.userName;
+                else
+                {
+                    model.recipientName = GetEllipsisFormat(message.recipient);
+                    model.isLoadingNames = true;
+
+                    // sometimes there is no cached profile, so we request it
+                    // dont block the operation of showing the message immediately
+                    // just update the message information after we get the profile
+                    EnsureProfileThenUpdateMessage(message.recipient, message, setScrollPositionToBottom, spamFiltering,
+                            limitMaxEntries, fetchProfileTries,
+                            profileFetchingCancellationToken.Token)
+                       .Forget();
+                }
+            }
+        }
+
+        if (!string.IsNullOrEmpty(message.sender))
+        {
+            model.senderId = message.sender;
+            UserProfile senderProfile = userProfileBridge.Get(message.sender);
+
+            if (senderProfile != null)
+                model.senderName = senderProfile.userName;
+            else
+            {
+                model.senderName = GetEllipsisFormat(message.sender);
+                model.isLoadingNames = true;
+
+                // sometimes there is no cached profile, so we request it
+                // dont block the operation of showing the message immediately
+                // just update the message information after we get the profile
+                EnsureProfileThenUpdateMessage(message.sender, message, setScrollPositionToBottom, spamFiltering,
+                        limitMaxEntries, fetchProfileTries,
+                        profileFetchingCancellationToken.Token)
+                   .Forget();
+            }
+        }
+
+        if (message.messageType == ChatMessage.Type.PRIVATE)
+        {
+            model.subType = message.sender == ownProfile.userId
+                ? ChatEntryModel.SubType.SENT
+                : ChatEntryModel.SubType.RECEIVED;
+        }
+        else if (message.messageType == ChatMessage.Type.PUBLIC)
+        {
+            model.subType = message.sender == ownProfile.userId
+                ? ChatEntryModel.SubType.SENT
+                : ChatEntryModel.SubType.RECEIVED;
+        }
+
+        AddChatMessage(model, setScrollPositionToBottom, spamFiltering, limitMaxEntries,
```

Comment:
This recursive flow is hard to follow. I would like to clarify how it should work:
1. It is called `AddChatMessage` but when called multiple times it acts as `update`?
2. Why do we call right the same method from `EnsureProfileThenUpdateMessage` when we already have `model` created and the only thing we need to do is update `recipientName` ? It leads to additional allocations and logical misintentions.